searchlib/src/vespa/searchlib/features/internal_max_reduce_prod_join_feature.cpp

```diff
@@ -0,0 +1,226 @@
+        FeatureExecutor(),
+        _attribute(attribute),
+        _queryVector(queryVector)
+{
+    _queryVector.syncMap();
+}
+
+template <typename A, typename V>
+feature_t maxProduct(const A& array, size_t count, const V& query)
+{
+    feature_t val = -DBL_MAX;
+    for (size_t i = 0; i < count; ++i) {
+        typename IntegerVector::HashMap::const_iterator itr = query.getDimMap().find(array[i].value());
+        if (itr != query.getDimMap().end()) {
+            feature_t v = itr->second; // weight from attribute is assumed to be 1.0
+            if (v > val) val = v;
+        }
+    }
+    return val == -DBL_MAX ? 0.0 : val;
+}
+
+template <typename BaseType>
+void
+RawExecutor<BaseType>::execute(uint32_t docId)
+{
+    const AT *values(nullptr);
+    const A *iattr = dynamic_cast<const A *>(_attribute);
+    size_t count = iattr->getRawValues(docId, values);
+    outputs().set_number(0, maxProduct(values, count, _queryVector));
+}
+
+/**
+ * Executor when array can't be accessed directly
+ */
+template <typename BaseType>
+class BufferedExecutor : public RawExecutor<BaseType> {
+private:
+    WeightedIntegerContent _buffer;
+
+public:
+    BufferedExecutor(const IAttributeVector * attribute, const IntegerVector & queryVector);
+    void execute(uint32_t docId) override;
+};
+
+template <typename BaseType>
+BufferedExecutor<BaseType>::BufferedExecutor(const IAttributeVector *attribute, const IntegerVector &queryVector) :
+    RawExecutor<BaseType>(attribute, queryVector),
+    _buffer()
+{
+}
+
+
+template <typename BaseType>
+void
+BufferedExecutor<BaseType>::execute(uint32_t docId)
+{
+    _buffer.fill(*RawExecutor<BaseType>::_attribute, docId);
+    feature_t val = maxProduct(_buffer, _buffer.size(), RawExecutor<BaseType>::_queryVector);
+    RawExecutor<BaseType>::outputs().set_number(0, val);
+}
+
+/**
+ * Blueprint
+ */
+InternalMaxReduceProdJoinBlueprint::InternalMaxReduceProdJoinBlueprint() :
+        Blueprint("internalMaxReduceProdJoin")
+{
+}
+
+InternalMaxReduceProdJoinBlueprint::~InternalMaxReduceProdJoinBlueprint()
+{
+}
+
+void
+InternalMaxReduceProdJoinBlueprint::visitDumpFeatures(const IIndexEnvironment &,
+                                                           IDumpFeatureVisitor &) const
+{
+}
+
+Blueprint::UP
+InternalMaxReduceProdJoinBlueprint::createInstance() const
+{
+    return Blueprint::UP(new InternalMaxReduceProdJoinBlueprint());
+}
+
+ParameterDescriptions
+InternalMaxReduceProdJoinBlueprint::getDescriptions() const
+{
+    return ParameterDescriptions().desc().attribute(ParameterCollection::ARRAY).string();
+}
+
+bool
+InternalMaxReduceProdJoinBlueprint::setup(const IIndexEnvironment &env, const ParameterList &params)
+{
+    _attribute = params[0].getValue();
+    _query = params[1].getValue();
+    describeOutput("scalar", "Internal executor for optimized execution of reduce(join(A,Q,f(x,y)(x*y)),max)");
+    env.hintAttributeAccess(_attribute);
+    return true;
```

Comment:
This is the last place we can reject the expression rewriting (the point of no return). Any case supported by tensorFromLabels(attribute(foo)) must either be rejected here or supported in createExecutor.